extensions/authentication-passport/src/__tests__/acceptance/authentication-with-passport-strategy-oauth2-adapter.acceptance.ts

```diff
@@ -0,0 +1,322 @@
+//  |            |                                                              |  v           |
+//  |            |      <------------------------[8]-----------------------     | ****         |
+//  +------------+        LB App returns to browser the access token            +--------------+
+
+/**
+ * options to pass to the Passport Strategy
+ */
+const oauth2Options: StrategyOptions = {
+  clientID: '1111',
+  clientSecret: '1917e2b73a87fd0c9a92afab64f6c8d4',
+  // `redirect_uri`:
+  // 'passport-oauth2' takes care of sending the configured `callBackURL` setting as `redirect_uri`
+  // to the authorization server. This behaviour is inherited by all other oauth2 modules like facebook, google, etc
+  callbackURL: 'http://localhost:8080/auth/thirdparty/callback',
+  // 'authorizationURL' is used by 'passport-oauth2' to begin the authorization grant flow
+  authorizationURL: 'http://localhost:9000/oauth/dialog',
+  // `tokenURL` is used by 'passport-oauth2' to exchange the access code for an access token
+  // this exchange is taken care internally by 'passport-oauth2'
+  // when the `callbackURL` is invoked by the authorization server
+  tokenURL: 'http://localhost:9000/oauth/token',
+};
+
+/**
+ * verify function for the oauth2 strategy
+ * This function mocks a lookup against a user profile datastore
+ *
+ * @param accessToken
+ * @param refreshToken
+ * @param profile
+ * @param done
+ */
+const verify: VerifyFunction = function (
+  accessToken: string,
+  refreshToken: string,
+  userProfile: MyUser,
+  done: VerifyCallback,
+) {
+  userProfile.token = accessToken;
+  if (!userRepository.findById(userProfile.id)) {
+    userRepository.createExternalUser(userProfile);
+  }
+  return done(null, userProfile);
+};
+
+/**
+ * convert user info to user profile
+ * @param user
+ */
+const myUserProfileFactory: UserProfileFactory<MyUser> = function (
+  user: MyUser,
+): UserProfile {
+  const userProfile = {[securityId]: user.id, ...user};
+  return userProfile;
+};
+
+/**
+ * Login controller for third party oauth provider
+ *
+ * This creates an authentication endpoint for the third party oauth provider
+ *
+ * Two methods are expected
+ *
+ * 1. loginToThirdParty
+ *           i. an endpoint for api clients to login via a third party app
+ *          ii. the passport strategy identifies this call as a redirection to third party
+ *         iii. this endpoint redirects to the third party authorization url
+ *
+ * 2. thirdPartyCallBack
+ *           i. this is the callback for the thirdparty app
+ *          ii. on successful user login the third party calls this endpoint with an access code
+ *         iii. the passport oauth2 strategy exchanges the code for an access token
+ *          iv. the passport oauth2 strategy then calls the provided `verify()` function with the access token
+ */
+export class Oauth2Controller {
+  constructor() {}
+
+  // this configures the oauth2 strategy
+  @authenticate('oauth2')
```

Comment:
@deepakrkris related to my comment in []()
I think the scenario here is different than what's expected from our authentication system...user profile injection should happen here, which means after the oauth2 strategy is invoked, a user profile should be returned as the result of the authentication action. 
My confusion point is, why we have to handle the redirect then get the user profile?

My expectation would be, the user profile is injected at controller level instead of the endpoint level. So that when the controller functions are invoked, the user profile is ready to use, both `/thirdparty` and `/callback` can print it.

I am not very familiar with the newly created `redirectRoute` therefore having a hard time to illustrate which part should handle the callback redirection, the strategy? or the controller?

Reply:
To explain that quickly, business`Usecase` endpoints wont be configured with an`oauth2` strategy. The `OAuth2Controller` is configured with an oauth2 strategy so that it can begin the `Authorization Code Grant` dialog. Once that flow is completed, the web client would switch to other strategies like `session`  or `jwt`.

Reply:
The longer version :)  : 
`/thirdparty` route acts as an interface to begin the oauth2 dialog with the thirdparty and `/callback` is for the third-party login to redirect back to LoopBack app. Once `/callback` endpoint has the oauth token + the user profile, it can proceed in three different ways. 

1. It can create a browser session. This is the most popular one we use everyday to login to an app using facebook/google credentials. This means the client from there on would use the `passport-session` strategy to access other `Usecase` endpoints in the LoopBack App.

2. Or it can return the original oauth token from the third-party to the web client which can then call other `Usecase` endpoints using the token as a `Bearer`(`jwt` strategy)

3. Or it can create a new token with the same expiration time of the original thirdparty token (the `exp` field in that token says it) and send that to the web client. (`jwt` strategy)